Support sorting based on values from a dictionary.

Subv opened the pull request:
## What, How & Why?
This is a draft and a request for comments about the feature and its implementation, I figured the best way to ask for help was to open a PR.

Note that it is currently possible to *filter* by a value stored under a specific dictionary key (using the `dictionary['key']` syntax) but is impossible to *sort* by that same key, this PR aims to change that.

I added some tests and FIXME comments about what I'm unsure about. This is my first time actually touching Realm's core code :)

Sorting is implemented by storing the parent dictionary's ColKey and the specific dictionary key we want to sort by.

This is a work-in-progress, the performance characteristics of searching the dictionary for the desired key on every iteration of the sort predicate are unknown.

The dictionary ColKey and the desired sort key are currently stored inside the ColKey structure to avoid changing too much of the SortDescriptor interface.

Sorting by dictionary keys using keyPath is currently not implemented (but desired).

In the future this could be extended to support sorting on specific indexes of a collection, ie. sort('array[0]').

## Pain points
- ~Changing the ColKey structure to store the extra fields needed feels like a hack, and makes it way bigger than a single 64 bit integer, this probably affects the cost of copying it all throughout the codebase. Maybe it would be better to store this data directly inside the SortDescriptor?~ This data is now stored in DictionaryKey and SortableColumnKey.
- The sort currently gets the parent dictionary and queries it for the sort key on _every_ iteration of the Sorter predicate, I'm not sure how fast that is, but the answer is probably "not at all".
- ~Currently only dictionaries of integers/strings are supported, this could probably be extended to arbitrary-length paths.~ Following links through the dictionary keys is now supported, it was nice low-hanging fruit after the SortableColumnKey implementation.
- There is very little error-checking in this patch, I wanted to get it out there and see what reaction I get about the feature.

## Example scenario
When you have data in the Entity-Attribute-Value pattern with dynamic attributes and want to allow the users of your application to sort by any of the fields, it's more convenient to have a dictionary of attributes keyed by attribute id in your table and be able to filter and sort by that directly.
Currently the way to solve that is to extract your Attributes into another table, filter that table by AttributeId, and then sort it based on the AttributeValue, after that you can grab the parent objects by looping through the attribute objects and grabbing a backlink to the parent Entity. But this seems convoluted when all you want is a simple sort.

## Update 1
Following feedback I reworked this patch so that the ColKey structure is unchanged. A new DictionaryKey structure was added to hold the required information to reference the specific dictionary keys, and the SortDescriptor and Sorter now deal with a derived type of std::variant<ColKey, DictionaryKey> called SortableColumnKey that deals with dispatching the proper key-specific behavior depending on the key type.

<!-- Describe the changes and give some hints to guide your reviewers if possible. -->
<!-- Link to relevant issue this fixes -->

## ☑️ ToDos
* [ ] 📝 Changelog update
* [X] 🚦 Tests (or not relevant)

Subv commented:
@cla-bot check

jedelbo commented:
@Subv Thank you for your work. I absolutely get the motivation for doing this, but as you have found out, the current design of the sort mechanism is not exactly geared towards this. I will definitely avoid adding to the ColKey struct! A more viable solution could be to have `ColumnsDescriptor::m_column_keys` be a vector of ColKey/Mixed union type of thing. Whether this can be done in an elegant way, I am not sure.

Subv commented:
@jedelbo Thank you for the feedback!

I reworked the patch a little to use a custom type derived from std::variant called SortableColumnKey and modified the sort descriptor code so that it can work with this new type. The variant takes care of abstracting the key-type-specific behaviors using std::visit.

tgoyne commented:
This seems like a reasonable approach to me. `child_key` probably needs to be a `std::string` as SortDescriptors can outlive the scope in which they're created and are used on different threads. We can't use `std::variant` because it requires a higher iOS deployment target than we currently have, but `mpark::variant` is a drop-in replacement for it.

Subv commented:
@tgoyne Thanks! I included the change for std::string child keys and the std::variant -> mpark::variant change.

Subv commented:
Rebased the PR to fix the conflicts and use the Table::check_column function instead of Table::report_invalid_link

jedelbo commented:
@Subv Your recent updates is more in line with what I would expect. However, I think that using the `mpark::variant` framework here is a bit of an overkill in this case. Instead of asking you tome make another attempt, I have created a competing PR (#5311). It is building on your work, but simplifies it a bit. This new PR is still in draft as I have to make some performance measurements in order to check how much the changes matter. Also some work on the query parser is missing.

jedelbo commented:
Closing as this is superseded by #5311
